python/datashaper/datashaper/table_store.py

```diff
@@ -61,5 +67,118 @@ class TableContainer(Generic[T]):
     """A container for a table and its metadata."""
 
     table: Table
+
+    # TODO: is this used? should we use it?
     metadata: Optional[TableMetadata] = None
     context: Optional[T] = None
+
+
+class TableStore:
+    """A table store for storing and retrieving tables."""
+
+    @abstractmethod
+    def add(self, name: str, table: Table) -> None:
+        """Add a table to the store."""
+        pass
+
+    @abstractmethod
+    def get(self, name: str) -> TableContainer:
+        """Get a table from the store."""
+        pass
+
+    @abstractmethod
+    def remove(self, name: str) -> None:
+        """Remove a table from the store."""
+        pass
+
+    @abstractmethod
+    def list(self) -> List[str]:
+        """List the tables in the store."""
+        pass
+
+    @abstractmethod
+    def clear(self) -> None:
+        """Clear the store."""
+        pass
+
+
+class DefaultTableStore(TableStore):
+    """A default table store implementation."""
+
+    def __init__(self) -> None:
+        self._tables: Dict[str, TableContainer] = {}
+
+    def add(self, name: str, table: TableContainer) -> None:
+        """Add a table to the store."""
+        self._tables[name] = table
+
+    def get(self, name: str) -> TableContainer:
+        """Get a table from the store."""
+        return self._tables[name]
+
+    def remove(self, name: str) -> None:
+        """Remove a table from the store."""
+        del self._tables[name]
+
+    def list(self) -> List[str]:
+        """List the tables in the store."""
+        return list(self._tables.keys())
+
+    def clear(self) -> None:
+        """Clear the store."""
+        self._tables.clear()
+
+
+class DiskCacheTableStore(TableStore):
+    """A default table store implementation."""
+
+    def __init__(
+        self, maxsize: int = 128, persist: bool = False, verbose=False
+    ) -> None:
+        self._persist = persist
+        self._maxsize = maxsize
+        self._verbose = verbose
+
+    def __enter__(self) -> TableStore:
+        self._path = tempfile.mkdtemp()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if not self._persist:
+            self.clear()
+
+    def add(self, name: str, table: TableContainer) -> None:
+        """Add a table to the store."""
+        if self._verbose:
+            print(f"saving table {name} to disk")
+        table.table.to_parquet(f"{self._path}/{name}.parquet")
+
+    @cache
+    def _get_caching_function(self):
+        @lru_cache(maxsize=self._maxsize)
+        def get_table(name: str) -> TableContainer:
+            if self._verbose:
+                print(f"loading table {name} from disk")
+            return TableContainer(pd.read_parquet(f"{self._path}/{name}.parquet"))
+
+        return get_table
+
+    def get(self, name: str) -> TableContainer:
+        """Get a table from the store."""
+        return self._get_caching_function()(name)

```

Comment:
nit:
Could we hace the caching_function as a property, so you could so self._cache_get(name) ?